File: 1887-reduction-operations-to-make-the-array-elements-equal/1887-reduction-operations-to-make-the-array-elements-equal.cpp
class Solution {
public:
    int reductionOperations(vector<int>& a) {
        
        map< int, int,greater<int>> m;
        for( int x:a) m[x]++;
    
        int prev=0, ans=0 , n=m.size() ;
        
        for( auto& t:m){
            if(n==1) break;
            ans=ans+ (t.second+prev);
            prev+=t.second;
            n--;
        }
        return ans;
    }
};
/*
input = [1,2,2,2,3,3,3,4]
1
1 1+3
5 4+3
*/

File: 1887-reduction-operations-to-make-the-array-elements-equal/1887-reduction-operations-to-make-the-array-elements-equal.java
import java.util.*;
class Solution {
    Comparator<Integer> Desc = new Comparator<Integer>() {
        @Override 
        public int compare(Integer a, Integer b) {
            return b - a;
        }           
    };
    
    public int reductionOperations(int[] a) {
        SortedMap<Integer, Integer> m = new TreeMap<>(Desc);
        for (int x : a) {
            m.put(x, m.getOrDefault(x, 0) + 1);
        }

        int prev = 0, ans = 0;
        for (Map.Entry<Integer, Integer> entry : m.entrySet()) {
            int freq = entry.getValue();
            ans += (freq + prev);
            prev += freq;
        }

        // Subtract the last added frequency (to exclude the largest element)
        ans -= prev;

        return ans;
    }
}
